### PR TITLE
Update mesh to handle multiple polynomial orders

### PR DESCRIPTION
**Pull request by jkozdon**
Update the mesh interface to support multiple polynomial orders in the mesh
construction (derivative matrices, quadrature rules, metrics terms,
connectivities between grid degrees of freedom)

The main user facing changes are that the `DiscontinuousSpectralElementGrid`
type now contains tuples of arrays for the

  - derivative matrix `D`
  - quadrature weights `ω`
  - indefinite integral operator `Imat`

(one element of for the operator for each the reference element directions)

The keyword argument `polynomialorder` of the `DiscontinuousSpectralElementGrid`
constructor can now either be an

  - `Int` in which case all dimension have the same polynomial order
  - `Ntuple{dim, Int}` corresponding to the polynomial order for each reference
    dimension
  - `Ntuple{2, Int}` if the mesh dimensionality is `3`, in which case the first
    argument is taken to the two horizontal polynomial orders (`ξ1` and `ξ2`)
    and the second is taken to be the vertical polynomial order (`ξ3`)

When the connectivity is constructed if the given polynomial orders will not
result in a conformal mesh then an error is thrown; note this implies that for
the cubed sphere mesh the polynomial orders must be the same in both horizontal
directions.

The `sgeo` data structure uses suboptimal storage in that it is of size
`(_nsgeo, maximum(Nfp), nface, nelem)` where 

  - `_nsgeo` is the number of surface metric terms
  - `Nfp` is the number of face points on each face
    (`Nfp = prod(Nq) ./ Nq` where `Nq = N .+ 1`)
  - `nface` is the number of reference element faces
  - `nelem` is the number of elements

This storage is suboptimal since `Nfp` is not constant for all the faces. A more
optimal storage pattern might be `(_nsgeo, 2prod(Nfp), nelem)` but this will
make the indexing into the array more complicated; possible future optimization.
Elements of `sgeo` that are not valid are set to `NaN`

Note that the no where in the actual compute kernels use the multiple polynomial
orders, this PR only updates the mesh data structure itself. Places where the
same polynomial order in all dimension needed to be assumed due to changes from
this PR are marked with the comment
```
# XXX: Needs updating for multiple polynomial orders
```

### CODE REVIEW

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try-

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try-

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try-

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try-

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
bors try

**Comment by jkozdon** [comment]
Increasing Floating type in the PR messed up some tests. Was orthogonal to the PR, so pulled out and opened #1682

File: experiments/OceanBoxGCM/simple_box.jl [hunk 1]
```diff
@@ -66,7 +66,12 @@ function run_simple_box(
         config_simple_box(name, resolution, dimensions, problem; BC = BC)
 
     grid = driver_config.grid
-    vert_filter = CutoffFilter(grid, polynomialorder(grid) - 1)
+    # XXX: Needs updating for multiple polynomial orders
```
[the remark this answers is not in the document]
Comment (jkozdon) [reply experiments/OceanBoxGCM/simple_box.jl:69]: No. we don't actually handle multiple polynomial orders, the code below assumes a single polynomial order.

The comments
```julia
# XXX: Needs updating for multiple polynomial orders
```
should help guide / locate all the places in the code that will need to be fixed after #1653 and #1656 are merged.

[the remark this answers is not in the document]
Comment (skandalaCLIMA) [reply experiments/OceanBoxGCM/simple_box.jl:69]: > Is this a stale comment now?

Once the PR is successfully merged, we need to revise code at the comment locations to handle different polynomial orders in each of the (3) directions.

[the remark this answers is not in the document]
Comment (valeriabarra) [reply experiments/OceanBoxGCM/simple_box.jl:69]: Ah @jkozdon sorry! The PR title had me believe that that was what this PR was doing and the comment had puzzled me for a sec. I'll keep reading, thanks!

File: src/Numerics/Mesh/Grids.jl [hunk 2]
```diff
@@ -167,27 +173,34 @@ struct DiscontinuousSpectralElementGrid{
     "Array indicating if a degree of freedom (real or ghost) is active"
     activedofs
 
-    "1-D lvl weights on the device"
+    "1-D lvl weights on the device (one for each dimension)"
```
Comment (valeriabarra) [inline src/Numerics/Mesh/Grids.jl:176]: Is `lvl` an acronym used elsewhere? I might have missed it

Comment (jkozdon) [reply src/Numerics/Mesh/Grids.jl:176]: Thanks. Should have been lgl

**Comment by valeriabarra** [comment]
Other than the comment I left above, LGTM overall. Of course I wasn't able to parse/fully understand all details (especially on what the tests do), but overall looks good. And passes all tests locally for me, too.

File: test/Numerics/DGMethods/advection_diffusion/pseudo1D_heat_eqn.jl [hunk 3]
```diff
@@ -214,19 +214,19 @@ let
     expected_result[2, 2, Float32, EveryDirection] = 6.5721644e-5
     expected_result[2, 3, Float32, EveryDirection] = 3.280845e-6
     expected_result[2, 1, Float32, HorizontalDirection] = 0.020514594
-    expected_result[2, 2, Float32, HorizontalDirection] = 0.0005684704
+    expected_result[2, 2, Float32, HorizontalDirection] = 0.0005687050
```
[the remark this answers is not in the document]
Comment (jkozdon) [reply test/Numerics/DGMethods/advection_diffusion/pseudo1D_heat_eqn.jl:217]: Good questions / point. My guess is that it is rearranging of the floating point operations in `computemetric!` of `src/Numerics/Mesh/Metrics.jl`. But can dig into this a bit.

**Comment by jkozdon** [comment]
> Other than the comment I left above, LGTM overall. Of course I wasn't able to parse/fully understand all details (especially on what the tests do), but overall looks good. And passes all tests locally for me, too.

Thanks @valeriabarra

**Comment by jkozdon** [comment]
FYI: I've updated the PR description, removed the PR from draft, and this is ready to review.

**Comment by skandalaCLIMA** [comment]
> FYI: I've updated the PR description, removed the PR from draft, and this is ready to review.

The PR clears all the tests locally.

File: test/Numerics/Mesh/Metrics.jl [hunk 4]
```diff
@@ -1,7 +1,7 @@
 using ClimateMachine.Mesh.Elements
 using ClimateMachine.Mesh.Metrics
 using Test
-
+using Random: MersenneTwister
```
Comment (charleskawczynski) [inline test/Numerics/Mesh/Metrics.jl:4]: Should we add a random seed?

Comment (jkozdon) [reply test/Numerics/Mesh/Metrics.jl:4]: It's added on line 266

File: test/Numerics/DGMethods/advection_diffusion/pseudo1D_heat_eqn.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (jkozdon) [reply test/Numerics/DGMethods/advection_diffusion/pseudo1D_heat_eqn.jl:217]: Guessing that most of these numbers I had to change when trying to use different floating point types. Most of them can be reverts, though the number for `expected_result[3, 2, Float32, EveryDirection]` is wrong on master (an artifact of `JULIA_CLIMA_INTEGRATION_TESTING = true` not being part of the CI system).

Going to revert all the numbers and open a separate PR (#1685)

**Review by valeriabarra — APPROVED** [review]

**Review by skandalaCLIMA — APPROVED** [review]

**Review by thomasgibson — APPROVED** [review]
LGTM

**Pull request merged**